File: frontend/src/application/use-cases/bookings/create-booking.js
// frontend/src/application/use-cases/create-booking.js

/**
 * @class CreateBookingUseCase
 * @description Caso de uso para crear una nueva reserva.
 */
export class CreateBookingUseCase {
    /**
     * @param {BookingRepository} bookingRepository - El repositorio de reservas.
     */
    constructor(bookingRepository) {
        this.bookingRepository = bookingRepository;
    }

    /**
     * Ejecuta el caso de uso para crear una reserva.
     * @param {string} courtId - El ID de la cancha.
     * @param {string} startTime - La hora de inicio de la reserva (formato ISO 8601).
     * @param {string} endTime - La hora de fin de la reserva (formato ISO 8601).
     * @returns {Promise<object>} Los datos de la reserva creada.
     * @throws {Error} Si ocurre un error al crear la reserva.
     */
    async execute(bookingDetails) {
        // Aquí se podría añadir lógica de validación de negocio antes de llamar al repositorio
        // Por ejemplo, verificar si el usuario tiene permisos para reservar, etc.

        const { courtId, startDateTime, endDateTime, paymentPercentage } = bookingDetails;

        try {
            const bookingData = {
                court: parseInt(courtId), // Convertir courtId a entero
                start_time: startDateTime, // Usar startDateTime
                end_time: endDateTime, // Usar endDateTime
                payment_percentage: paymentPercentage, // Incluir el porcentaje de pago
            };
            return await this.bookingRepository.createBooking(bookingData);
        } catch (error) {
           // console.error("Error en CreateBookingUseCase:", error);
            throw new Error("No se pudo crear la reserva.");
        }
    }
}

// TODO: Definir la interfaz BookingRepository en frontend/src/domain/repositories/booking_repository.js
// TODO: Implementar ApiBookingRepository en frontend/src/infrastructure/repositories/api-booking-repository.js

File: backend/bookings/application/use_cases/create_booking.py
from typing import Dict, Any
from ...domain.repositories.booking_repository import IBookingRepository
from ...models import Booking # Asumiendo que Booking está en backend/bookings/models.py
from users.models import User # Para el tipo de usuario

class CreateBookingUseCase:
    """
    Caso de uso para crear una nueva reserva.
    Esta clase reside en la capa de Aplicación.
    """
    def __init__(self, booking_repository: IBookingRepository):
        self.booking_repository = booking_repository

    async def execute(self, booking_data: Dict[str, Any]) -> Booking:
        # Aquí se podría añadir lógica de aplicación adicional si fuera necesario
        # (ej. validaciones de negocio complejas, notificaciones, etc.)
        # La validación de disponibilidad ya se maneja en el repositorio por ahora.
        # El usuario ahora está incluido en booking_data
       # print("Datos de reserva enviados al repositorio:", booking_data) # Debug print
        return await self.booking_repository.create(booking_data)
